puf_data/impute_itmexp.py

```diff
@@ -0,0 +1,321 @@
+represent the correlations between the several itemized expense
+variables.  This method is sometimes called sequential regression
+multiple imputation.  See Raghunathan, et al., "A Multivariate
+Technique for Multiply Imputing Missing Values Using a Sequence of
+Regression Models" (2001).
+
+Imputing amounts for nonitemizers using statistical models estimated
+on itemizers requires that we handle the resulting Heckman sample
+selection problem: any statistical model of the itemizers that is used
+to impute itemized expense amounts for nonitemizers will over-estimate
+the imputed amounts.  This problem is handled by using three different
+ad hoc procedures to handle the Heckman sample selection problem.
+(Numbered comments below contain more detail on these three procedures.)
+
+And one additional procedure is used in this work: we scale the
+distribution of each itemized expense variable so that the weighted
+count of nonitemizers with a positive imputed amount and the weighted
+dollar sum of the imputed amounts approximate those estimated by JCT
+in JCX-75-15, "Estimating Changes in the Federal Individual Income
+Tax: Description of the Individual Tax Model," April 23, 2015, pages
+18-22, as summarized in Table 2 on page 22, which is entitled "Number
+of Tax Filing Units and Amounts of Imputed Itemized Deductions for
+Non-Itemizers, 2011."  (Comments below contain more detail on this
+procedure.)
+"""
+from __future__ import print_function
+import numpy as np
+import pandas as pd
+import statsmodels.api as sm
+
+
+DUMP0 = False
+DUMP1 = False
+DUMP2 = False
+CALIBRATING = False
+
+
+def impute(ievar, logit_prob_af, log_amount_af,
+           exogenous_vars, itemizer_data, nonitemizer_data):
+    """
+    Function that estimates imputation equations for ievar with itemizer_data
+    using the list of exogenous variables.  The estimated equations are then
+    used (along with the two additive factors) to impute amounts for ievar
+    for nonitemizers with the imputed nonitemizer amounts being returned.
+    """
+    if DUMP1:
+        print('****** IMPUTE {} ******'.format(ievar))
+    # estimate Logit parameters for probability of having a positive amount
+    logit_y = (itemizer_data[ievar] > 0).astype(int)
+    logit_x = itemizer_data[exogenous_vars]
+    logit_res = sm.Logit(logit_y, logit_x).fit(disp=0)
+    x_b = logit_res.predict(nonitemizer_data[exogenous_vars], linear=True)
+    exp_x_b = np.exp(x_b + logit_prob_af[ievar])
+    adj_prob = exp_x_b / (1.0 + exp_x_b)
+    np.random.seed(int(ievar[1:]))
+    urn = np.random.uniform(size=len(x_b))
+    positive_imputed = np.where(urn < adj_prob, True, False)
+    if DUMP1:
+        print(logit_res.summary())
+        print(adj_prob.head())
+        print(round(positive_imputed.mean(), 4))
+        print(len(nonitemizer_data))
+    # estimate OLS parameters for the positive amount using a sample of
+    # itemizers who have positive ievar amounts than are less than the
```

Comment:
Did you mean "...amounts *that* are..."?

Reply:
Yes, thanks for catching this mistake.